### PR TITLE
TensorGraph GraphConvs

### PR DESCRIPTION
**Pull request by lilleswing**
Decided To Finish GraphConvs before working on the TensorGraph API Changes

Results as a spreadsheet of the algorithm.
https://docs.google.com/spreadsheets/d/1rCNT47BiEHpTfENa-1KTupjEWCH0hraP6rYatEiaQUA/edit#gid=958335762

Jenkins Job Calculating Results
http://bellsprout:8080/job/tensorgraph_benchmarks/10/console

Code will change after new TensorGraph API changes, but layers will probably stay the same.

### CODE REVIEW

**Comment by peastman** [comment]
> Jenkins Job Calculating Results
> http://bellsprout:8080/job/tensorgraph_benchmarks/10/console

I get a broken link when I click on it.

**Comment by lilleswing** [comment]
@peastman sorry -- from inside stanford VPN
http://171.65.103.54:8080/job/tensorgraph_benchmarks/10/console

The role username and password are on tensorbro in the lab.

We have to come up for a way to expose this resource more fully before I leave.

**Comment by lilleswing** [comment]
P.S. @peastman the biggest "gotcha" that I found re-implementing this was that dc.nn.copy.Dense does NOT actually use the "activation" arg.  So they are all linear.  Making my architecture reproduce always using linear instead of the different args passed in gave a significant boost to my performance.

**Comment by lilleswing** [comment]
Also of note -- there are still some perf issues

1) tf.dynamic_partition is CPU only.  This is such a big deal that tf generally just does everything on the CPU instead of sliding things in and out of the GPU (at least for smaller networks).  This is used in GraphGather to turn atoms back into molecules.  We can probably come up with something that only uses GPU ops.

2) Queues are not in use because batches are molecule based, but n molecules can have different number of atoms, all inputs into queues must have fixed sizes.   We can probably solve this by padding our batches and then have an initial layer clear off the padding, but I'm leaving that to be done.

File: deepchem/models/tensorgraph/models/graph_conv.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+class GraphConvTensorGraph(TensorGraph):
+  """
+  """
+
+  def __init__(self, **kwargs):
+    super().__init__(**kwargs)
+    self.min_degree = 0
+    self.max_degree = 10
+
+  def _construct_feed_dict(self, X_b, y_b, w_b, ids_b):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_conv.py]: This looks totally fine for this PR, but I think we should make a generic solution to constructing feed-dicts for graphs with many inputs (DTNNs, Weaves, DAGs, etc all have many different input types).  Ideally, all the MoleculeNet models should require no custom subclasses with methods beyond those in `TensorGraph`

I have an idea for a general solution. Will post on #498.

File: deepchem/models/tensorgraph/models/graph_conv.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            print('Ending epoch %d: Average loss %g' % (epoch, avg_loss))
+        saver.save(sess, self.save_file, global_step=self.global_step)
+        self.last_checkpoint = saver.last_checkpoints[-1]
+      ############################################################## TIMING
+      time2 = time.time()
+      print("TIMING: model fitting took %0.3f s" % (time2 - time1))
+      ############################################################## TIMING
+
+
+def graph_conv_model(batch_size, num_tasks):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_conv.py]: Should we define a default graph-conv model? I can see how it would make sense for new users, but I also like the idea of showing users that it's not magic under the hood. After the API refactor, these models should be straightforward to define.

File: examples/tox21/tox21_graph_conv.py [hunk 3]
```diff
@@ -20,35 +21,13 @@
...
-    batch_size=batch_size,
-    learning_rate=1e-3,
-    learning_rate_decay_time=1000,
-    optimizer_type="adam",
-    beta1=.9,
-    beta2=.999)
-
-# Fit trained model
-model.fit(train_dataset, nb_epoch=10)
+
```
Comment (rbharath) [inline examples/tox21/tox21_graph_conv.py]: Couple requests here:

1. Let's move the `TensorGraph` graph-conv example to a new example file. We have a number of users of the existing graph-convs, so we should do a slow phase-out of the previous graph-convs over the next couple version bumps.
2. For the example, let's explicitly define the graph convs here instead of using the pre-built function.

**Comment by rbharath** [comment]
@lilleswing Actually, slightly amending my review comment, could you restore the old Graph-conv example before the merge?

**Comment by lilleswing** [comment]
@rbharath yeah I will restore the old one before the merge and also in benchmark.py.  

After the full run we aren't getting as good results as before.

https://docs.google.com/spreadsheets/d/1rCNT47BiEHpTfENa-1KTupjEWCH0hraP6rYatEiaQUA/edit#gid=958335762

File: deepchem/models/tensorgraph/models/graph_conv.py (hunk 2 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/models/graph_conv.py]: I kinda like having an example which has been shown to do well on lots of cases.  Reading through this function is pretty easy.

File: examples/tox21/tox21_graph_conv.py (hunk 3 above)
Comment (lilleswing) [inline examples/tox21/tox21_graph_conv.py]: Sure we can do that -- I think there is value in showing that there is a single network that does well on all of them though.

File: deepchem/models/tensorgraph/models/graph_conv.py (hunk 1 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/models/graph_conv.py]: yeah I think the API suggested in #498 should work for multi-input.

Comment (lilleswing) [inline deepchem/models/tensorgraph/models/graph_conv.py]: however for graph convs the atom molecule membership is based off the batch and calculated at runtime based on the batch.  This api won't be able to handle this special case.

Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_conv.py]: Good point! I have a thought for a slight modification to the API that should catch this case. Will make an update to #498

File: examples/tox21/tox21_graph_conv.py (hunk 3 above)
Comment (rbharath) [inline examples/tox21/tox21_graph_conv.py]: Definitely like the idea of showing that a single network does well on many inputs. Perhaps two examples then? One with the default network and one with a custom network (for beginner and intermediate users)

**Comment by lilleswing** [comment]
@rbharath restored all old code -- further API changes can be added via the tensorgraph API changes

File: examples/tox21/tox21_graph_conv.py (hunk 3 above)
Comment (rbharath) [inline examples/tox21/tox21_graph_conv.py]: It looks like the original example is still deleted?

**Comment by rbharath** [comment]
Looks like `tox21_graph_conv.py` is still modified? Good to merge in once that's reverted (or moved to a different file)

**Comment by lilleswing** [comment]
tox21_graph_conv is back

**Comment by rbharath** [comment]
@lilleswing Awesome, thanks for the fix!
